#### ДЗ к уроку 2. Дискретные случайные величины. Закон распределения вероятностей. Биномиальный закон распределения. Распределение Пуассона

##### Задача 1
Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

##### Решение:  

1) Результат ответа на любой из вопросов не зависит от ответов на другие вопросы, т.е. события независимые.    
2) Вероятность правильного ответа на любой из вопросов равна $$p = \frac{1}{4}$$  
3) Исходя из п.1. и п.2 распределение является Биномиальным.  Поэтому вероятности вычисляются по формуле Бернулли:  
$$P_n(X=k) = C_n^k p^k q^{n-k}$$  
где n = 5 - количеству вопросов,  
k - количество верно отвеченных вопросов,  
$q = 1 - p =  \frac{3}{4}$

$$P(X=0) = C_5^0 \cdot (\frac{1}{4})^0 \cdot (\frac{3}{4})^5=(\frac{3}{4})^5=0.2373046875$$  

$$P(X=1) = C_5^1 \cdot (\frac{1}{4})^1 \cdot (\frac{3}{4})^4 = 0.3955078125$$  

$$P(X=2) = C_5^2 \cdot (\frac{1}{4})^2 \cdot (\frac{3}{4})^3=0.263671875$$  

$$P(X=3) = C_5^3 \cdot (\frac{1}{4})^3 \cdot (\frac{3}{4})^2=0.087890625$$  

$$P(X=4) = C_5^4 \cdot (\frac{1}{4})^4 \cdot (\frac{3}{4})^1=0.0146484375$$  

$$P(X=5) = C_5^5 \cdot (\frac{1}{4})^5 \cdot (\frac{3}{4})^0=0.0009765625$$  

##### Ответ:  
Математическое ожидание и дисперсия биномиального распределения:  
$$M(X) = np = 5 \cdot \frac{1}{4} = 1.25$$  
$$D(X) = npq = 5 \cdot \frac{1}{4} \cdot \frac{3}{4} = 0.9375$$

In [1]:
import numpy as np
N = 5
PA = 0.25

def combinations(n, k):
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

def P(x):
    return combinations(N, x)*(PA)**x*(1 - PA)**(N-x)

def M(x, p):
    spam = 0
    for i in range(len(x)):
        spam += x[i]*p[i]
    return spam

def D(x, p):
    m = M(x, p)
    return  M((x - m)**2,p)

def init_p(X):
    p  =[]
    for i in X:
        p.append(P(i))
    return p

def print_p(X, p, start, end):
    for i in X:
        if i>=start and i<=end:
            print(f'P(X={i})={p[i]}')

X = np.arange(N+1)
p = init_p(X)
print_p(X,p,0, len(X)-1)
print(f'Сумма всех вероятностей:{np.sum(p)}')
m = M(X, p)
print(f'Матожидание:{m}')
d = D(X, p)
print(f'Дисперсия:{d}')

P(X=0)=0.2373046875
P(X=1)=0.3955078125
P(X=2)=0.263671875
P(X=3)=0.087890625
P(X=4)=0.0146484375
P(X=5)=0.0009765625
Сумма всех вероятностей:1.0
Матожидание:1.25
Дисперсия:0.9375


##### Задача 2  

Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

##### Решение:

1) Результат одного выстрела не влияет на результат другого, т.е. события независимые.    
2) Вероятность попадания для любого выстрела равна $p = 0.01$  
3) Исходя из п.1. и п.2 распределение является Биномиальным.  Поэтому вероятности вычисляются по формуле Бернулли:  
$$P(X=k) = C_n^k p^k q^{n-k}$$  
где n = 200 - количеству выстрелов,  
k - количество попаданий,  
q = 1 - p = 0.99.  
Вероятность попадания не менее 5 и не более 10 равна сумме вероятностей событий("только 5 попаданий из 200", "только 6 попаданий из 200",..., "только 10 попаданий из 200"), т.к. события несовместны. 

##### Ответ:

$$P(X=5) = C_{200}^{5} \cdot (0,01)^5 \cdot (0.99)^{195}=0.03572335672160829$$  

$$P(X=6) = C_{200}^{6} \cdot (0,01)^6 \cdot (0.99)^{194}=0.011727364580325954$$  

$$P(X=7) = C_{200}^{7} \cdot (0,01)^7 \cdot (0.99)^{193}=0.0032829851783307863$$  

$$P(X=8) = C_{200}^{8} \cdot (0,01)^8 \cdot (0.99)^{192}=0.0008000203780528306$$  

$$P(X=9) = C_{200}^{9} \cdot (0,01)^9 \cdot (0.99)^{191}=0.000172394963620812$$ 

$$P(X=10) = C_{200}^{10} \cdot (0,01)^{10} \cdot (0.99)^{190}=3.326003843593443e-05$$  

$$P(5<=X<=10) = P(X=5) + P(X=6) + P(X=7) + P(X=8) + P(X=9) + P(X=10)= 0.051739381860374606$$ 

In [2]:
N = 200 # количество выстрелов
PA = 0.01 # вероятность попадания
min_goal=5 # минимальное количество попаданий
max_goal=10 # максимальное количество попаданий
X = np.arange(N+1) # события 
p = init_p(X)
print_p(X,p,min_goal,max_goal)
print(f'Сумма вероятностей попаданий не менее {min_goal} и не более {max_goal} равна :{np.sum(p[min_goal:max_goal+1])}')

P(X=5)=0.03572335672160829
P(X=6)=0.011727364580325954
P(X=7)=0.0032829851783307863
P(X=8)=0.0008000203780528306
P(X=9)=0.000172394963620812
P(X=10)=3.326003843593443e-05
Сумма вероятностей попаданий не менее 5 и не более 10 равна :0.051739381860374606


##### Задача 3

В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?

Подсказка. Как посчитать P(X >= k) для величины X, имеющей распределение Пуассона? Здесь предстоит немножко покодить.

##### Решение:

Вероятность попадания не менее 10 раз P(X >= 10) равна 1 - P(X < 10)  
Формула Пуассона:
$$P_m \approx \frac{\lambda^m}{m!}e^{-\lambda}$$  
$λ = np$, перебирая n будем искать вероятность не ниже 0.9

In [5]:
from scipy import optimize

min_goal = 10# минимальное количество попаданий
PA = 0.01 # вероятност попадания
tol=1e-15 # точность оптимизации
n = 200 # начальное значения количества выстрелов

def P_more(n):
    spam = 0
    lambda_ = n * PA
    for i in range(min_goal):
         spam += (lambda_**i / np.math.factorial(i)) * np.exp(-lambda_)
    return 1 - spam 

def f(n, tol = tol):
    return (P_more(n) - 0.9 - tol)**2

result = optimize.minimize(f, n, tol = tol)
x = np.round(result.x[0])
print(f'При {x-1} вероятность попаданий не менее 10 раз равна:{P_more(x-1)}')
print(f'При {x} вероятность попаданий не менее 10 раз равна:{P_more(x)}')
print(f'При {x+1} вероятность попаданий не менее 10 раз равна:{P_more(x+1)}')

При 1420.0 вероятность попаданий не менее 10 раз равна:0.8997364748565644
При 1421.0 вероятность попаданий не менее 10 раз равна:0.9001760722984115
При 1422.0 вероятность попаданий не менее 10 раз равна:0.9006140609380022


##### Ответ:
Чтобы быть с вероятностью 0.9, тобы хотя бы 10 раз будет совершено попадание, необходимо сделать не менее 1421 выстрел